From 002d20ade8d60d940f8722b1ffea433a671f345a Mon Sep 17 00:00:00 2001
From: Jonathon Jongsma <jjongsma@redhat.com>
Date: Thu, 25 Sep 2014 10:53:55 -0500
Subject: [PATCH] Unset app 'fullscreen' when leaving fullscreen

Previously, the fullscreen floating toolbar and the "toggle-fullscreen"
hotkey (which maps to the menu item action) had slightly different
methods of exiting fullscreen. The floating toolbar method unset the
'fullscreen' property on the application (which causes all windows to
simultaneously exit fullscreen), whereas the hotkey did not. This had a
side-effect of preventing the display from auto re-sizing if it was
fullscreened again.  After this change, both the hotkey and the toolbar
button will unset the application-level 'fullscreen' property when
exiting fullscreen mode.

Resolves: rhbz#1022608
(cherry picked from commit c4d1347a6d9893c34bb50c3fafcd7c9c23e1dc08)
---
 src/virt-viewer-window.c | 28 ++++++++++++++++++----------
 1 file changed, 18 insertions(+), 10 deletions(-)

diff --git a/src/virt-viewer-window.c b/src/virt-viewer-window.c
index c8d5722..8dbf5a5 100644
--- a/src/virt-viewer-window.c
+++ b/src/virt-viewer-window.c
@@ -830,15 +830,26 @@ virt_viewer_window_menu_file_quit(GtkWidget *src G_GNUC_UNUSED,
 
 
 static void
+virt_viewer_window_set_fullscreen(VirtViewerWindow *self,
+                                  gboolean fullscreen)
+{
+    if (fullscreen) {
+        virt_viewer_window_enter_fullscreen(self, -1);
+    } else {
+        /* leave all windows fullscreen state */
+        if (virt_viewer_app_get_fullscreen(self->priv->app))
+            g_object_set(self->priv->app, "fullscreen", FALSE, NULL);
+        /* or just this window */
+        else
+            virt_viewer_window_leave_fullscreen(self);
+    }
+}
+
+static void
 virt_viewer_window_toolbar_leave_fullscreen(GtkWidget *button G_GNUC_UNUSED,
                                             VirtViewerWindow *self)
 {
-    /* leave all windows fullscreen state */
-    if (virt_viewer_app_get_fullscreen(self->priv->app))
-        g_object_set(self->priv->app, "fullscreen", FALSE, NULL);
-    /* or just this window */
-    else
-        virt_viewer_window_leave_fullscreen(self);
+    virt_viewer_window_set_fullscreen(self, FALSE);
 }
 
 static void keycombo_menu_location(GtkMenu *menu G_GNUC_UNUSED, gint *x, gint *y,
@@ -875,10 +886,7 @@ virt_viewer_window_menu_view_fullscreen(GtkWidget *menu,
 {
     gboolean fullscreen = gtk_check_menu_item_get_active(GTK_CHECK_MENU_ITEM(menu));
 
-    if (fullscreen)
-        virt_viewer_window_enter_fullscreen(self, -1);
-    else
-        virt_viewer_window_leave_fullscreen(self);
+    virt_viewer_window_set_fullscreen(self, fullscreen);
 }
 
 static void add_if_writable (GdkPixbufFormat *data, GHashTable *formats)
